### PR TITLE
[WIP] Proposal to enable webpack bundle optimization and html customization

### PR DESCRIPTION
**Pull request by davidfestal**
### Pull Request Goal

The goal of this PR is to provide a way to:
- optionally optimize the packaging of bundles to make them caching-compliant, following some of the guidelines provided [here](https://webpack.js.org/guides/caching/).
- customize the generated `index.html`, by optionally providing a custom template to be used by the [Html Webpack Plugin](https://webpack.js.org/plugins/html-webpack-plugin/) and specifying additional template parameters. 

A typical use of these proposed features is to allow retrieving some of the generated bundle chunks from a CDN, and fallbacking to the Theia server when CDN files are not available.
 
#### Bundle optimization

Relevant [Webpack optimization options](https://webpack.js.org/configuration/optimization/) are added when the `--env-optimized` parameter is passed to the build.

#### HTML customization

- The [Html Webpack Plugin](https://webpack.js.org/plugins/html-webpack-plugin/) is now used in all cases, though producing the same output as before in the default case
- If a `custom-html-head.html` file is present in the Webpack build directory, its content is used inside the `head` tag of the Html template, instead of the default behavior that simply adds the `script` tags required to load the Javascript bundles (as was before).
In other words, when optimization is disabled, the default `head` tag of the generated Html would look like:
```
<head>
  <meta charset="UTF-8">
  <script src="bundle.js" type="text/javascript" charset="utf-8"></script>
</head>
``` 
But when the following `custom-html-head.html` file would be in the build folder:
```
  <script src="<%= htmlWebpackPlugin.options.someOptionsToProvideCDNURL %>" type="text/javascript" charset="utf-8"></script>
```
then the default `script` tag would be replaced by a line generated from this custom head template.
- The `--env--customtemplateparams` command-line build argument allows providing additional template parameters to be used by the `custom-html-head.html` template fragment, as a JSON string like this:
```
--env--customtemplateparams='{ "cdnPrefix": "some CDN URL" }'
```
Then, inside the template, this value can be used with the following code:
```
<%= htmlWebpackPlugin.options.customparams.cdnPrefix %>
```
 
#### Remarks

- This PR goes in the same direction as PR https://github.com/theia-ide/theia/pull/3398/files and seems to be somewhat a continuation of it.

### CODE REVIEW

**Comment by davidfestal** [comment]
cc @benoitf @uniibu

**Comment by uniibu** [comment]
This line [#L69](https://github.com/davidfestal/theia/blob/544f3a000f7ff2e352df6ce36998398d25a4c86e/dev-packages/application-manager/src/application-package-manager.ts#L69) should be removed to avoid any conflict with `html-webpack-plugin` and besides, this pr renders that function useless.

Also instead of using custom templates for use with like cdn services, why not just let webpack do the generation as well. This would be less prone to dev error.
Something like:
```js
    output: {
        filename: optimized ? '[name].[contenthash].js' : 'bundle.js',
        path: optimized ? path.resolve(__dirname, 'cdn/assets') : outputPath,
	publicPath: optimized ? cdnPrefix : '' // cdnPrefix would be like https://cdn.example.com/assets/
    }
```
Then have `html-webpack-plugin` inject it on the head/body. Only caveat is that the dev will not be able to add custom external/remote script, but i think that is ideal for lesser prone dev error.

**Comment by davidfestal** [comment]
@uniibu 

> This line #L69 should be removed to avoid any conflict with html-webpack-plugin and besides, this pr renders that function useless.

Sure, I'll change this.

> Then have html-webpack-plugin inject it on the head/body.

well it doesn't seem flexible enough to allow:
- retrieving *only some of the files* from an alternate location
- retrieving the file from the Theia web server when it is not available on the CDN

**Comment by uniibu** [comment]
> There are 2 modes: production and development. In prod mode, Theia should be built optimized. What is the difference of `--env-optimized` to building in the production mode.
> 

I think the reasoning for this is to make this PR optional, as stated by @geropl on the other [PR](https://github.com/theia-ide/theia/pull/3398#issuecomment-436529563), this might cause issues to the infrastructure where Theia is deployed to.

File: dev-packages/application-manager/src/generator/webpack-generator.ts [hunk 1]
```diff
@@ -67,6 +90,7 @@ module.exports = {
         net: 'empty',
         crypto: 'empty'`)}
     },
+    optimization: optimizeOptions,
```
[the remark this answers is not in the document]
Comment (davidfestal) [reply dev-packages/application-manager/src/generator/webpack-generator.ts]: very interesting option. I'll see if/how I can use it to replace the webpack generator part of this PR.

**Comment by davidfestal** [comment]
> > There are 2 modes: production and development. In prod mode, Theia should be built optimized. What is the difference of `--env-optimized` to building in the production mode.
> 
> I think the reasoning for this is to make this PR optional, as stated by @geropl on the other [PR](https://github.com/theia-ide/theia/pull/3398#issuecomment-436529563), this might cause issues to the infrastructure where Theia is deployed to.

Exactly, the idea is that without this option the produced files will be the same as before, and nothing would be broken for existing users.
But I could check to disallow this option in development mode if you want @akosyakov

**Comment by davidfestal** [comment]
@akosyakov 

>If you do optimization, please bring numbers before/after for connections of different quality. One should be able to reproduce these numbers on his own to confirm that PR is a good optimization to go in the prod mode. If it is not good, but you still want to have it then use a custom webpack config for your app.

> Particularly to code splitting, we've tried it and it performed worse on slow connection than with one chunk.

Well there would be only 3 files there instead of one (I mean that's not *many* files).
I might also be that using `async` / `defer` options in the `script` tags might help leveraging concurrent downloads in the case of code splitting.

Anyway, this option is mainly a pre-requisite (or at least good-to-have) for long-term caching,  as explained in the mentioned Webpack documentation page, and has not much sense if caching is not planned. This is also why I made it optional. I could change the name of the option to `cacheable`, or something like that if it makes it clearer.

**Comment by akosyakov** [comment]
my point is there is already a way to add optional webpack configuration for its app, explained here: https://github.com/theia-ide/theia/issues/766#issuecomment-341720642

I don't see the point to add optional flags which are disabled by default next to it. If there is something which we can enable by default for prod/dev mode and everybody will for sure benefit from it, then it should be introduced and an optional flag should be added to disable it.

**Comment by uniibu** [comment]
> I don't see the point to add optional flags which are disabled by default next to it. If there is something which we can enable by default for prod/dev mode and everybody will for sure benefit from it, then it should be introduced and an optional flag should be added to disable it.

I do agree that if there is something which everyone/most could benefit,it should be introduced. However, based on a [comment](https://github.com/theia-ide/theia/pull/3398#issuecomment-436529563) on the previous/linked pr, Theia tends to lean more towards business/vendor usage/deployment. It is why this pr, even if it could benefit most, must be set optionally and disabled by default.

> my point is there is already a way to add optional webpack configuration for a Theia app, explained here: #766 (comment)

This is true for the most part. However customizations such as the goal of this PR can't be achieve solely by a custom webpack config. For one, this PR's goal for making a `caching-compliant` assets would be useless unless there is a way(correct me if im wrong) for a dev to set options to `express.static`(coz it always responds with `max-age:0`) without having to build a separate extension just for that purpose.

**Comment by uniibu** [comment]
Did some tests in regards to the time and speed.

**Using my own pc with cache disabled:**

Current(single bunde.js): load time of bundle.js **13.66s**
pccurrent

Optimized(chunked): total load time of all chunked js **8.86s**
pcchunked

**Using GTMetrix(online)**

Current(single bundle.js): load time of bundle.js **8.68s**
gtcurrent

Optimized(chunked): total load time of all chunked js **4.15s**
gtchunked

I did not add the test for `--split-frontend` or development mode as that takes too much time and latency to load.

The difference of this pr vs the current(single bundle.js) and `--split-frontend` is that the js files are already in-lined in the html file which means browsers can take advantage of parallel connections. Unlike the current or the `--split-frontend`, the js files are loaded programmatically via single js file,which doesn't take advantage of parallel connections and blocks rendering.

**Comment by davidfestal** [comment]
Thanks @uniibu for these tests, I'm not surprised of the results since I've been testing a POC on my side as well these last days.

**Comment by davidfestal** [comment]
> For one, this PR's goal for making a caching-compliant assets would be useless unless there is a way(correct me if im wrong) for a dev to set options to express.static(coz it always responds with max-age:0) without having to build a separate extension just for that purpose

@uniibu unless the caching-compliant files are then pushed to a CDN and retrieved from there through a customized `index.html` page, which is the solution I'm working on.

**Comment by davidfestal** [comment]
@akosyakov I'll try to use an additional webpack configuration on our side to avoid the changes in the Theia webpack configurator.

But are you OK with the html-related changes (use a custom head template if found in the build directory) ?

**Comment by vince-fugnitto** [comment]
@davidfestal do you still plan on updating this pull-request, the changes are quite old and there has been no recent activity. It would be good to keep the list of open pull-requests to a manageable number.

**Comment by vince-fugnitto** [comment]
> @davidfestal do you still plan on updating this pull-request, the changes are quite old and there has been no recent activity. It would be good to keep the list of open pull-requests to a manageable number.

Closing due to inactivity. Please feel free to re-open the pull-request whenever it is updated.

**Pull request closed without merging**